pubnub-gson/pubnub-gson-api/src/main/java/com/pubnub/api/java/models/consumer/objects_api/membership/PNChannelMembership.java

```diff
@@ -20,31 +20,80 @@ public static class ChannelId {
 
     @Getter
     private final ChannelId channel;
+    @Getter
+    private final Map<String, Object> custom;
+    @Getter
+    private final String status;
+    @Getter
+    private final String type;
 
+    @Deprecated
     public static PNChannelMembership channel(final String channelId) {
         return new JustChannel(new ChannelId(channelId));
     }
 
+    @Deprecated
     public static PNChannelMembership channelWithCustom(final String channelId, final Map<String, Object> custom) {
-        return new ChannelWithCustom(new ChannelId(channelId), new HashMap<>(custom));
+        return new ChannelWithCustom(new ChannelId(channelId), new HashMap<>(custom), null, null);
+    }
+
+    // PNChannelMembership.PNChannelMembership(membership01ChannelId).type("t").status("s").custom(custom).build()
+    // add static import to get PNChannelMembership(membership01ChannelId).type("t").status("s").custom(custom).build()
+    @SuppressWarnings("checkstyle:MethodName")
+    public static Builder PNChannelMembership(final String channelId) {
+        return new Builder(new ChannelId(channelId));
+    }
+
+    public static class Builder {
+        private final ChannelId channelId;
+        private Map<String, Object> custom = new HashMap<>();
+        private String status;
+        private String type;
+
+        private Builder(ChannelId channelId) {
+            this.channelId = channelId;
+        }
+
+        public Builder custom(Map<String, Object> custom) {
+            this.custom = new HashMap<>(custom);
+            return this;
+        }
+
+        public Builder status(String status) {
+            this.status = status;
+            return this;
+        }
+
+        public Builder type(String type) {
+            this.type = type;
+            return this;
+        }
+
+        public PNChannelMembership build() {
+            if (custom.isEmpty() && status == null && type == null) {
+                return new JustChannel(channelId);
+            } else {
+                return new ChannelWithCustom(channelId, custom, status, type);
+            }
+        }
     }
 
     @Getter
     @EqualsAndHashCode(callSuper = true)
     public static class JustChannel extends PNChannelMembership {
         JustChannel(@NonNull final ChannelId channelId) {
-            super(channelId);
+            super(channelId, null, null, null);
         }
     }
 
     @Getter
     @EqualsAndHashCode(callSuper = true)
     public static class ChannelWithCustom extends PNChannelMembership {
-        private final Object custom;
-
-        ChannelWithCustom(@NonNull final ChannelId channelId, @NonNull Object custom) {
-            super(channelId);
-            this.custom = custom;
+        ChannelWithCustom(@NonNull final ChannelId channelId,
+                          @NonNull Map<String, Object> custom,
+                          String status,
```

Comment:
this is a breaking change, leave it as it was, and in `public PNChannelMembership build()` just return an instance of PNChannelMembership:

You can try like this (I don't remember how Java works) :
`return new PNChannelMembership(channelId, custom, status, type) {};`

if that doesn't work maybe just make PNChannelMembership non-abstract or add a 
`private static class PNChannelMembershipImpl extends PNChannelMembership`

Reply:
ok